### ISSUE TITLE
Extending schema to change properties of existing keys

### ISSUE THREAD

**Issue by mprager** [alecthomas/voluptuous#219]
Currently modifying key types in an existing schema duplicates the keys and causes all objects to fail validation. Consider this example:

```
>>> s1 = Schema({Optional('a'): int})
>>> s2 = s1.extend({Required('a'): int})
>>> s1.schema
{'a': <type 'int'>}
>>> s2.schema
{'a': <type 'int'>, 'a': <type 'int'>}
```

The s2 object now cannot pass validation for any object, since all 'a' keys will be considered for the optional key and the required key appears absent.

```
>>> s1({})
{}
>>> s1({'a': 3})
{'a': 3}
>>> s2({})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/mprager/.virtualenv/validator/lib/python2.7/site-packages/voluptuous/schema_builder.py", line 192, in __call__
    return self._compiled([], data)
  File "/home/mprager/.virtualenv/validator/lib/python2.7/site-packages/voluptuous/schema_builder.py", line 486, in validate_dict
    return base_validate(path, iteritems(data), out)
  File "/home/mprager/.virtualenv/validator/lib/python2.7/site-packages/voluptuous/schema_builder.py", line 324, in validate_mapping
    raise er.MultipleInvalid(errors)
voluptuous.error.MultipleInvalid: required key not provided @ data['a']
>>> s2({'a':4})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/mprager/.virtualenv/validator/lib/python2.7/site-packages/voluptuous/schema_builder.py", line 192, in __call__
    return self._compiled([], data)
  File "/home/mprager/.virtualenv/validator/lib/python2.7/site-packages/voluptuous/schema_builder.py", line 486, in validate_dict
    return base_validate(path, iteritems(data), out)
  File "/home/mprager/.virtualenv/validator/lib/python2.7/site-packages/voluptuous/schema_builder.py", line 324, in validate_mapping
    raise er.MultipleInvalid(errors)
voluptuous.error.MultipleInvalid: required key not provided @ data['a']
```

**Comment by tusharmakkar08** [comment]
Hey @mprager 

Thanks for pointing out the issue. This looks like a bug in voluptuous. Are you interested in submitting a PR for fixing this ? I can point you to the place where you need to investigate this issue. 

Thanks.

**Comment by mprager** [comment]
Hey sure thing.

I have some changes I made locally tot he extend function in the Schema
class. They seem to work, I'll pull and submit the changes.

Mike

**Comment by tusharmakkar08** [comment]
Closed via 99fb7bd43b2cbd8e53cb9e50765c93339b2f7843
